network/discovery/src/lib.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::counters::DISCOVERY_COUNTS;
+use async_trait::async_trait;
+use channel::diem_channel::Receiver;
+use diem_config::{config::PeerSet, network_id::NetworkContext};
+use diem_logger::prelude::*;
+use diem_types::on_chain_config::{OnChainConfigPayload, ON_CHAIN_CONFIG_REGISTRY};
+use futures::{Stream, StreamExt};
+use network::{
+    connectivity_manager::{ConnectivityRequest, DiscoverySource},
+    counters::inc_by_with_context,
+    logging::NetworkSchema,
+};
+use std::pin::Pin;
+use subscription_service::ReconfigSubscription;
+use tokio::runtime::Handle;
+
+mod counters;
+mod file;
+mod validator_set;
+
+pub use file::FileChangeListener;
+pub use validator_set::ValidatorSetChangeListener;
+
+/// A union type for all implementations of `DiscoveryChangeListenerTrait`
+pub enum DiscoveryChangeListener {
+    ValidatorSet(ValidatorSetChangeListener),
+    File(FileChangeListener),
+}
+
+impl DiscoveryChangeListener {
+    pub fn start(self, executor: &Handle) {
+        match self {
+            Self::ValidatorSet(listener) => executor.spawn(Box::pin(listener).start()),
+            Self::File(listener) => executor.spawn(Box::pin(listener).start()),
+        };
+    }
+
+    pub fn discovery_source(&self) -> DiscoverySource {
+        match self {
+            DiscoveryChangeListener::ValidatorSet(listener) => listener.discovery_source(),
+            DiscoveryChangeListener::File(listener) => listener.discovery_source(),
+        }
+    }
+}
+
+/// A listener for changes on an incoming discovery source.  The entire set of discovered peers must
+/// be provided in each update.
+#[async_trait]
+pub trait DiscoveryChangeListenerTrait: Stream<Item = PeerSet> {
+    /// `NetworkContext` for logging
+    fn network_context(&self) -> &NetworkContext;
+
+    /// The `DiscoverySource` for purposes of keeping track of each source separately
+    fn discovery_source(&self) -> DiscoverySource;
+
+    /// The `channel::Sender` for updating networking with a new `PeerSet`
+    fn update_channel(&self) -> channel::Sender<ConnectivityRequest>;
```

Comment:
you can make this `&mut self -> &mut Sender` just need to add `Unpin` trait bound to `DiscoveryChangeListenerTrait`

Reply:
I can't add `Unpin` to `DiscoveryChangeListenerTrait` because of the `Sleep` or `Interval`, which is how it ended up like this.